apps/scratch/src/utilities/student-activity-tracking.ts

```diff
@@ -0,0 +1,91 @@
+import { StudentAction } from "../types/scratch-student-activities";
+import { StudentActivityHandlerParams } from "../types/scratch-student-activities";
+import { processStudentActivityPipeline } from "./scratch-student-activities/pipeline";
+import { mapXmlBlockToBlock } from "./scratch-block";
+import type { WorkspaceChangeEvent } from "../types/scratch-workspace";
+
+const scratchToStudentAction: Record<string, StudentAction> = {
+  create: StudentAction.Create,
+  move: StudentAction.Move,
+  delete: StudentAction.Delete,
+};
+
+const TRACKED_ACTIONS = [
+  StudentAction.Create,
+  StudentAction.Move,
+  StudentAction.Delete,
+] as const;
+
+export const mapScratchEventTypeToStudentAction = (
+  type: string,
+): StudentAction | null => scratchToStudentAction[type] || null;
+
+export const shouldRecordStudentAction = (
+  action: StudentAction,
+  event: WorkspaceChangeEvent,
+  canEditTask: boolean | undefined,
+): boolean | undefined =>
+  // This condition ensures that we only track user-initiated block moves or creations
+  // when undo is enabled, the task is not editable, and the block has an ID.
```

Comment:
When `undo` is enabled, the task is *not* editable? Let's have a chat about this one